WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -77,44 +53,22 @@ private CoinsView AsSpentCoinsView()
 
 	public bool TryAdd(SmartCoin coin)
 	{
-		var added = false;
 		lock (Lock)
 		{
-			if (!SpentCoins.Contains(coin))
+			coin.RegisterToHdPubKey();
+
+			if (Transactions.TryGetValue(coin.TransactionId, out var transactionInternal))
 			{
-				added = Coins.Add(coin);
-				coin.RegisterToHdPubKey();
-				if (added)
-				{
-					foreach (var outPoint in coin.Transaction.Transaction.Inputs.Select(x => x.PrevOut))
-					{
-						var newCoinSet = new HashSet<SmartCoin> { coin };
-
-						// If we don't succeed to add a new entry to the dictionary.
-						if (!CoinsByOutPoint.TryAdd(outPoint, newCoinSet))
-						{
-							var previousCoinTxId = CoinsByOutPoint[outPoint].First().TransactionId;
-
-							// Then check if we're in the same transaction as the previous coins in the dictionary are.
-							if (coin.TransactionId == previousCoinTxId)
-							{
-								// If we are in the same transaction, then just add it to value set.
-								CoinsByOutPoint[outPoint].Add(coin);
-							}
-							else
-							{
-								// If we aren't in the same transaction, then it's a conflict, so replace the old set with the new one.
-								CoinsByOutPoint[outPoint] = newCoinSet;
-							}
-						}
-					}
-
-					InvalidateSnapshot = true;
-				}
+				return transactionInternal.Coins.TryAdd(coin.Outpoint, coin);
 			}
-		}
 
-		return added;
+			Transactions.Add(
+				coin.TransactionId,
+				new TransactionInternal(
+					new Dictionary<OutPoint, SmartCoin>() {{ coin.Outpoint, coin }},
+					coin.Transaction.Transaction.Inputs.Select(x => x.PrevOut).ToHashSet()));
+		}
+		return true;
```

Comment:
Here is the idea of #11384: I group this cache by `Transaction`. I don't check anymore if there is a conflict because anyway if an input of a transaction changes it will be replaced and removed from `Transactions`.